packages/backend-core/src/middleware/passport/sso/sso.ts

```diff
@@ -0,0 +1,163 @@
+import { generateGlobalUserID } from "../../../db"
+import { authError } from "../utils"
+import * as users from "../../../users"
+import * as context from "../../../context"
+import fetch from "node-fetch"
+import { SSOAuthDetails, SSOUser, User } from "@budibase/types"
+
+type SaveUserOpts = {
+  requirePassword?: boolean
+  hashPassword?: boolean
+  currentUserId?: string
+}
+
+export type SaveUserFunction = (
+  user: SSOUser,
+  opts: SaveUserOpts
+) => Promise<any>
+
+/**
+ * Common authentication logic for third parties. e.g. OAuth, OIDC.
+ */
+export async function authenticate(
+  details: SSOAuthDetails,
+  requireLocalAccount: boolean = true,
+  done: any,
+  saveUserFn?: SaveUserFunction
```

Comment:
This should not allow `undefined` if then we throw an error. This should be typed as required and use to the validation in build time

Reply:
Good call on this - it's likely the source of a bug with datasource auth and user token refresh - updated in: https://github.com/Budibase/budibase/pull/9672/commits/a8ab00ca0eba9cc3e13d7de170d8a64d43ab0aa5